sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala

```diff
@@ -232,7 +233,19 @@ class V2SessionCatalog(catalog: SessionCatalog)
         throw QueryCompilationErrors.tableAlreadyExistsError(ident)
     }
 
-    loadTable(ident)
+    val table = loadTable(ident)
+
+    // Check if the schema of the created table matches the given schema.
+    if (schema.nonEmpty) {
+      val tableSchema = CharVarcharUtils.replaceCharVarcharWithStringInSchema(
+        table.columns().asSchema)
+      if (!DataType.equalsIgnoreNullability(tableSchema, schema)) {
```

Comment:
can we check it in `loadTable`? In case the data source is a bit random and only return wrong schema at second time.

Reply:
I tried it but looks like `loadTable` is used in many other places, such as in `tableExists`. If we check the schema there, it will fail commands other than create table (e.g `DROP TABLE t` will fail because the schema in catalogTable does not match the table schema)